Improves workflow for provisioned dashboards

bergquist opened the pull request:
Closes #10883

![image](https://user-images.githubusercontent.com/618863/37969793-f8eed3a2-31d1-11e8-9a77-b67752669ebf.png)

bergquist committed:
provisioning: improve UX when saving provisioned dashboards

bergquist committed:
dashboards: reject updates of provisioned dashboards

bergquist committed:
provisioning: better description for provisioned save modal

torkelo commented:
Nice !

torkelo commented:
Maybe we can use the code editor and have indented json instead like view json views? (not sure if it is, maybe it just looks unformatted).

marefr commented:
How about something like this? 

![image](https://user-images.githubusercontent.com/1668778/38383349-9a5d1320-390c-11e8-85da-98ee1e7ca2f3.png)

Added a save to file as well since I think that could be a good complement.

torkelo commented:
Nice!

marefr committed:
provisioning: ux fixes when saving provisioned dashboards

Wider modal.
Tried to improve provisioning description text.
Code editor for json.
Button that allows to save the json to a file.

marefr committed:
Merge branch 'master' into readonly_dashboards

marefr committed:
provisioning: fix tests for save provisioned dashboard modal

bergquist commented:
Great work! :) anything else we want to do before we merge this?

marefr commented on pkg/api/dashboard.go:
```diff
@@ -102,6 +102,13 @@ func GetDashboard(c *m.ReqContext) Response {
+	dpQuery := &m.GetProvisionedDashboardByDashboardId{DashboardId: dash.Id}
+	err = bus.Dispatch(dpQuery)
+	if dpQuery.Result != nil {
+		meta.CanEdit = true
```
Can overriding *CanEdit* create any issues with permissions?

marefr commented on pkg/api/dashboard.go:
```diff
@@ -102,6 +102,13 @@ func GetDashboard(c *m.ReqContext) Response {
 	}
 
+	dpQuery := &m.GetProvisionedDashboardByDashboardId{DashboardId: dash.Id}
+	err = bus.Dispatch(dpQuery)
```
*err* doesn't seem to be handled.

bergquist replied:
good catch!

marefr commented on pkg/models/dashboards.go:
```diff
@@ -13,26 +13,28 @@ import (
+	ErrDashboardUpdateAccessDenied             = errors.New("Access denied to save dashboard")
+	ErrDashboardInvalidUid                     = errors.New("uid contains illegal characters")
+	ErrDashboardUidToLong                      = errors.New("uid to long. max 40 characters")
+	ErrDashboardCannotSaveProvisionedDashboard = errors.New("Cannot save provisioned dashboards")
```
Should be *Cannot save provisioned dashboard*

marefr commented on pkg/services/dashboards/dashboard_service.go:
```diff
@@ -111,6 +111,11 @@ func (dr *dashboardServiceImpl) buildSaveDashboardCommand(dto *SaveDashboardDTO,
 		return nil, models.ErrDashboardUpdateAccessDenied
 	}
 
+	err := dr.validateDashboardIsNotProvisioned(dash.Id)
```
This should probably be done before ValidateDashboardBeforeSaveCommand since the dashboardId can be changed in that validation if title changed for example.

marefr commented on pkg/services/sqlstore/dashboard_provisioning.go:
```diff
@@ -17,6 +18,23 @@ type DashboardExtras struct {
+func GetProvisionedDataByDashboardId(cmd *models.GetProvisionedDashboardByDashboardId) error {
+	result := &models.DashboardProvisioning{}
+
+	exist, err := x.Where("dashboard_id = ?", cmd.DashboardId).Get(result)
```
Is orgId required as well or maybe I'm missing something?

bergquist replied:
That table does not have OrgID at the moment. Since all dashboards have their own Id, regardless of org it's not really needed.

marefr replied:
Cool :+1:

marefr commented on pkg/services/sqlstore/dashboard_provisioning.go:
```diff
@@ -17,6 +18,23 @@ type DashboardExtras struct {
+	}
+
+	if !exist {
+		return models.ErrDashboardProvisioningDoesNotExist
```
A bit confused. Why not return nil in this case or use the cmd.Result to tell if it's provisioned or not?

bergquist replied:
Using errors for signaling that something expected was not found is quite common in go. 
ex https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/user.go#L158
So for this query I think it makes sense since it returns the ProvisionedDashboard struct.

We could create a new query to check if it exists or not but I felt that was kinda redundant.

marefr commented on pkg/services/dashboards/dashboard_service.go:
```diff
@@ -129,6 +134,23 @@ func (dr *dashboardServiceImpl) buildSaveDashboardCommand(dto *SaveDashboardDTO,
+	}
+
+	if err == nil {
+		return models.ErrDashboardCannotSaveProvisionedDashboard
```
Logic here would probably be easier to read if GetProvisionedDashboardByDashboardId handler returns err for actual error and set the cmd.Result. Then cmd.Result can be used to decide if error message should be returned or not.

marefr reviewed (changes requested):
@bergquist Hadn't reviewed the code before so did that now since you asked about the state of PR.

Added some comments that we maybe can discuss.

bergquist committed:
provisioning: adds error handling

bergquist committed:
provisioning: fixes typo

bergquist committed:
provisioning: check provisiong before saveCmd

bergquist committed:
provisioning: fixes broken tests

bergquist committed:
provisioning: simplify db query

the GetProvisionedDashboardQuery wasent used for anything
else than check if a dashboard is provisioned or not. So
we simplified this query to make it more maintainable.

bergquist committed:
provisioning: dont override permissions

bergquist commented:
ready for review again :)

marefr commented on pkg/models/dashboards.go:
```diff
@@ -32,9 +32,9 @@ var (
-	ErrDashboardProvisioningDoesNotExist       = errors.New("Dashboard provisioning does not exist")
-	RootFolderName                             = "General"
+	ErrDashboardCannotSaveProvisionedDashboard = errors.New("Cannot save provisioned dashboard")
+	//ErrDashboardProvisioningDoesNotExist       = errors.New("Dashboard provisioning does not exist")
```
Remove

marefr commented on pkg/services/sqlstore/dashboard_provisioning.go:
```diff
@@ -18,19 +18,15 @@ type DashboardExtras struct {
-	}
-
-	cmd.Result = result
+	cmd.Result = exist
```
Nice :+1:

marefr commented on pkg/api/dashboard.go:
```diff
@@ -102,10 +102,13 @@ func GetDashboard(c *m.ReqContext) Response {
+	isDashboardProvisioned := &m.IsDashboardProvisionedQuery{DashboardId: dash.Id}
+	err = bus.Dispatch(isDashboardProvisioned)
+	if err != nil {
+		return Error(500, "Error while checking if dashboard is provisioned", err)
```
Nice :+1:

marefr reviewed (changes requested):
Looks good :+1: Just a comment that should be removed. 

Haven't had time to take this for a test spin yet. WIll do that tomorrow.

marefr committed:
Merge branch 'master' into readonly_dashboards

marefr committed:
remove comment/unused variable

marefr commented on pkg/services/dashboards/dashboard_service.go:
```diff
@@ -93,6 +93,17 @@ func (dr *dashboardServiceImpl) buildSaveDashboardCommand(dto *SaveDashboardDTO,
+		return nil, err
+	}
+
+	if isDashboardProvisioned.Result {
```
This basically don't allow any provisioned dashboards to be saved

marefr reviewed (changes requested):
Critical bug that we need to fix

marefr committed:
provisioned dashboard validation should not be made from provisioning service

Provisioned dashboard validation should not be made when creating/updating a folder or
when provisioning service are trying to provision/create/update dashboard

marefr committed:
provisioned dashboard validation should be made when importing a dashboard

Had to add the validate before save command before validating provisioned dashboard
so that the import would get a proper dashboard id

marefr commented:
@bergquist could you please review my latest 2 commits. It's basically the second one I'm most concerned about since I switched order of when the dashboard provisioning validation are done. Had to do the validate before save to get a proper dashboard id for the provisioning validation.

marefr commented on pkg/services/dashboards/dashboard_service.go:
```diff
@@ -93,6 +93,16 @@ func (dr *dashboardServiceImpl) buildSaveDashboardCommand(dto *SaveDashboardDTO,
+		Overwrite: dto.Overwrite,
+	}
+
+	if err := bus.Dispatch(&validateBeforeSaveCmd); err != nil {
```
I can see one case where we may have a problem with doing the validate before save. If validate before save returns a `ErrDashboardVersionMismatch` or `ErrDashboardWithSameNameInFolderExists` the result would be a overwrite warning dialog in UI. If user clicks on overwrite it will popup "you cannot save provisioned dashboard" dialog if the dashboard is provisioned. 

Unsure if this is acceptable or if we need to add some special case handling here? Guess it depends on what use case we're trying to solve.

bergquist replied:
This could only happen when users try to save over a provisioned dashboard. Right?

If that's the case we should give the user `You cannot save provisioned dashboard` dialog.

marefr replied:
> This could only happen when users try to save over a provisioned dashboard. Right?

Yes

> If that's the case we should give the user You cannot save provisioned dashboard dialog.

Then we need to add a special case that we should validate provisioned dashboard if `err = ErrDashboardVersionMismatch` or `err = ErrDashboardWithSameNameInFolderExists`. Will add some complexity, but absolutely doable. What do you think?

bergquist replied:
So.... :) 

I think we might want to check twice if the dashboard is provisioned? before validateSave and after? since the id might update during validateSave.

marefr replied:
That will not solve the problem with `ErrDashboardVersionMismatch` or `ErrDashboardWithSameNameInFolderExists`. At the same time it will not allow you to overwrite another dashboard that's not provisioned by updating title of a provisioned dashboard.

marefr commented:
@bergquist sorry for bugging you, but would need some feedback on this to be able to hopefully finish and merge it for v5.1.

bergquist commented on pkg/services/dashboards/dashboard_service.go:
```diff
@@ -57,7 +57,7 @@ func (dr *dashboardServiceImpl) GetProvisionedDashboardData(name string) ([]*mod
 }
 
-func (dr *dashboardServiceImpl) buildSaveDashboardCommand(dto *SaveDashboardDTO, validateAlerts bool) (*models.SaveDashboardCommand, error) {
+func (dr *dashboardServiceImpl) buildSaveDashboardCommand(dto *SaveDashboardDTO, validateAlerts bool, validateProvisionedDashboard bool) (*models.SaveDashboardCommand, error) {
```
One bool parameter can be ok. When we have two we should refactor :) But let's do that post merge.

marefr committed:
Merge branch 'master' into readonly_dashboards

marefr committed:
revert changes of add panel button to require save permission

This resolves an issue where the add panel button was shown on the default home dashboard

marefr committed:
Merge branch 'master' into readonly_dashboards
